modules/core/keeper/msg_server.go

```diff
@@ -751,7 +751,101 @@ func (k Keeper) ChannelUpgradeInit(goCtx context.Context, msg *channeltypes.MsgC
+
+	module, _, err := k.ChannelKeeper.LookupModuleByChannel(ctx, msg.PortId, msg.ChannelId)
+	if err != nil {
+		ctx.Logger().Error("channel upgrade try failed", "port-id", msg.PortId, "error", errorsmod.Wrap(err, "could not retrieve module from port-id"))
+		return nil, errorsmod.Wrap(err, "could not retrieve module from port-id")
+	}
+
+	cbs, ok := k.Router.GetRoute(module)
+	if !ok {
+		ctx.Logger().Error("channel upgrade try failed", "port-id", msg.PortId, "error", errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module))
+		return nil, errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module)
+	}
+
+	proposedUpgrade, expectedCounterpartyChannel, err := k.ChannelKeeper.ChanUpgradeTry(
+		ctx,
+		msg.PortId,
+		msg.ChannelId,
+		msg.ProposedUpgradeConnectionHops,
+		msg.ProposedUpgradeTimeout,
+		msg.CounterpartyProposedUpgrade,
+		msg.CounterpartyUpgradeSequence,
+		msg.ProofChannel,
+		msg.ProofUpgrade,
+		msg.ProofHeight,
+	)
+	if err != nil {
+		ctx.Logger().Error("channel upgrade try failed", "error", errorsmod.Wrap(err, "channel handshake upgrade try failed"))
+		return &channeltypes.MsgChannelUpgradeTryResponse{
+			ChannelId: msg.ChannelId,
+			Version:   proposedUpgrade.Fields.Version,
+			Result:    channeltypes.FAILURE,
+		}, err
+	}
+
+	// call blockUpgrade handshake to move channel from INITUPGRADE to BLOCKUPGRADE
+	if err := k.ChannelKeeper.BlockUpgradeHandshake(
+		ctx,
+		msg.PortId,
+		msg.ChannelId,
+		proposedUpgrade.Fields,
+		expectedCounterpartyChannel,
+		msg.CounterpartyProposedUpgrade,
+		msg.ProofChannel,
+		msg.ProofUpgrade,
+		msg.ProofHeight,
+	); err != nil {
+		ctx.Logger().Error("channel upgrade try failed", "error", errorsmod.Wrap(err, "channel handshake upgrade try failed"))
+		return &channeltypes.MsgChannelUpgradeTryResponse{
+			ChannelId: msg.ChannelId,
+			Version:   proposedUpgrade.Fields.Version,
+			Result:    channeltypes.FAILURE,
+		}, err
+	}
+
+	proposedVersion, err := cbs.OnChanUpgradeTry(
+		ctx,
+		msg.PortId,
+		msg.ChannelId,
+		msg.CounterpartyProposedUpgrade.Fields.Ordering,
+		msg.CounterpartyProposedUpgrade.Fields.ConnectionHops,
+		msg.CounterpartyProposedUpgrade.Fields.Version,
+	)
+	if err != nil {
+		ctx.Logger().Error("channel upgrade try callback failed", "port-id", msg.PortId, "channel-id", msg.ChannelId, "error", err.Error())
+		// TODO: commit error receipt to state and abort channel upgrade
+		return &channeltypes.MsgChannelUpgradeTryResponse{
+			ChannelId: msg.ChannelId,
+			Version:   proposedVersion,
+			Result:    channeltypes.FAILURE,
+		}, err
+	}
+
+	// set version to return value of callback
+	proposedUpgrade.Fields.Version = proposedVersion
+
+	// grab channel her to get latest version after UpgradeTry
+	channel, found := k.ChannelKeeper.GetChannel(ctx, msg.PortId, msg.ChannelId)
+	if !found {
+		return nil, errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "port ID (%s) channel ID (%s)", msg.PortId, msg.ChannelId)
+	}
```

Comment:
Should we maybe just do this get inside `WriteUpgradeTryChannel`? This way we can never accidentally pass in a stale channel to the function.

Reply:
could make sense...but then WriteUpgradeTry is doing much more than just writing the upgrade try, and we should change the method name

Reply:
Lets move into `WriteUpgradeTry`. This mimics how we do it in the opening handshake. I'm perfectly happy for `WriteUpgradeTry` to get the channel in order to update it's state